internal/server/filesystem/server.go

```diff
@@ -2,24 +2,227 @@ package filesystem
+	if pathCtx == internal.PLUGIN {
+		prefix = s.kubeletCSIPluginsPath
+	} else if pathCtx == internal.CONTAINER {
+		prefix = s.kubeletPodPath
+	} else {
+		return fmt.Errorf("Invalid PathContext: %v", pathCtx)
+	}
+
+	pathlen := len(path)
+
+	if pathlen > MAX_PATH_LENGTH_WINDOWS {
+		return fmt.Errorf("Path length %d exceeds maximum characters: %d", pathlen, MAX_PATH_LENGTH_WINDOWS)
+	}
+
+	if len(path) > 0 && (path[0] == '\\') {
+		return fmt.Errorf("Invalid character \\ at begining of path: %s", path)
+	}
+
+	if isUNCPathWindows(path) {
+		return fmt.Errorf("Unsupported UNC path prefix: %s", path)
+	}
+
+	if containsInvalidCharactersWindows(path) {
+		return fmt.Errorf("Path contains invalid characters: %s", path)
+	}
+
+	if isAbsWindows(path) && !strings.HasPrefix(path, prefix) {
+		return fmt.Errorf("Absolute path: %s is not within context path: %s", path, prefix)
+	}
+
+	return nil
+}
+
+func (s *Server) abs(pathCtx internal.PathContext, path string) (string, error) {
+	if isAbsWindows(path) {
+		return path, nil
+	}
+	prefix := ""
+	if pathCtx == internal.PLUGIN {
+		prefix = s.kubeletCSIPluginsPath
+	} else if pathCtx == internal.CONTAINER {
+		prefix = s.kubeletPodPath
+	} else {
+		return "", fmt.Errorf("Invalid PathContext: %v", pathCtx)
+	}
+	return prefix + "\\" + path, nil
+}
 
 // PathExists checks if the given path exists on the host.
 func (s *Server) PathExists(ctx context.Context, request *internal.PathExistsRequest, version apiversion.Version) (*internal.PathExistsResponse, error) {
-	// FIXME: actually implement this!
+	err := s.validatePathWindows(request.Context, request.Path)
+	if err != nil {
+		return &internal.PathExistsResponse{
+			Error: err.Error(),
+		}, nil
+	}
+
+	path, err := s.abs(request.Context, request.Path)
+	if err != nil {
+		return &internal.PathExistsResponse{
+			Error: err.Error(),
+		}, nil
+	}
+
+	exists, err := s.hostAPI.PathExists(path)
 	return &internal.PathExistsResponse{
-		Success: false,
-		CmdletError: &api.CmdletError{
-			CmdletName: "dummy",
-			Code:       12,
-			Message:    "hey there " + request.Path,
-		},
+		Error:  err.Error(),
+		Exists: exists,
+	}, nil
+}
+
+func (s *Server) Mkdir(ctx context.Context, request *internal.MkdirRequest, version apiversion.Version) (*internal.MkdirResponse, error) {
+	err := s.validatePathWindows(request.Context, request.Path)
+	if err != nil {
+		return &internal.MkdirResponse{
+			Error: err.Error(),
+		}, nil
+	}
+
+	path, err := s.abs(request.Context, request.Path)
+	if err != nil {
+		return &internal.MkdirResponse{
+			Error: err.Error(),
+		}, nil
+	}
+
+	err = s.hostAPI.Mkdir(path)
+	if err != nil {
+		return &internal.MkdirResponse{
+			Error: err.Error(),
+		}, nil
+	}
+
+	return &internal.MkdirResponse{
+		Error: "",
+	}, nil
+}
+
+func (s *Server) Rmdir(ctx context.Context, request *internal.RmdirRequest, version apiversion.Version) (*internal.RmdirResponse, error) {
+	err := s.validatePathWindows(request.Context, request.Path)
```

Comment:
the passing parameter ctx context.Context is not used?

Reply:
Both are intentional. For `ctx`, please see [above](https://github.com/kubernetes-csi/csi-proxy/pull/11#discussion_r339925937). For `apiversion.Version`, it is not required right now given this is the v1alpha1 - the initial implementation. It will become relevant as we add more versions of the APIs and there arises a need to differentiate between API versions that are sending the calls to the underlying/internal implementation after parameter conversions.